atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/knowledge/coordinated/CoordinationAwareKnownAbandonedTransactionsStore.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * (c) Copyright 2022 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.transaction.knowledge.coordinated;
+
+import com.google.common.collect.Iterables;
+import com.google.common.collect.RangeMap;
+import com.palantir.atlasdb.coordination.CoordinationService;
+import com.palantir.atlasdb.coordination.ValueAndBound;
+import com.palantir.atlasdb.internalschema.InternalSchemaMetadata;
+import com.palantir.atlasdb.transaction.impl.TransactionConstants;
+import com.palantir.atlasdb.transaction.knowledge.AbandonedTimestampStore;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public final class CoordinationAwareKnownAbandonedTransactionsStore {
+    private final CoordinationService<InternalSchemaMetadata> coordinationService;
+    private final AbandonedTimestampStore delegate;
+
+    public CoordinationAwareKnownAbandonedTransactionsStore(
+            CoordinationService<InternalSchemaMetadata> coordinationService, AbandonedTimestampStore delegate) {
+        this.coordinationService = coordinationService;
+        this.delegate = delegate;
+    }
+
+    public void addAbandonedTimestamps(Set<Long> abandonedTimestamps) {
+        List<Long> sortedTs = abandonedTimestamps.stream().sorted().collect(Collectors.toList());
+        Long greatestAbandonedTs = Iterables.getLast(sortedTs);
+        RangeMap<Long, Integer> transactionsSchemaMap = latestTimestampRangesSnapshot(greatestAbandonedTs);
+        Set<Long> abandonedTsOnSchema4 = sortedTs.stream()
+                .filter(ts -> {
+                    Optional<Integer> maybeSchema = Optional.ofNullable(transactionsSchemaMap.get(ts));
+                    return maybeSchema
+                            .map(integer -> integer.equals(TransactionConstants.TTS_TRANSACTIONS_SCHEMA_VERSION))
```

Comment:
interesting, i suppose we could do that, except this operation is actually triggered by sweep. So it would not really matter if we are up/downgrading. We will always be writing to A.